src/ripple/net/ShardDownloader.md

```diff
@@ -0,0 +1,263 @@
+
+```c++
+void SSLHTTPDownloader::do_session()
+{
+
+   // (Connection initialization logic)
+
+   .
+   .
+   .
+
+   // (In between calls to async_read_some):
+   if(isStopped_.load())
+   {
+       close(p);
+       return exit();
+   }
+
+   .
+   .
+   .
+
+   break;
+}
+```
+
+### Recovery
+
+Persisting the current state of both the archive handler and the downloader is achieved by leveraging a SQLite database rather than flat files, as the database protects against data corruption that could result from a system crash.
+
+##### ShardArchiveHandler
+
+Although `SSLHTTPDownloader` is a generic class that could be used to download a variety of file types, currently it is used exclusively by the `ShardArchiveHandler` to download shards. In order to provide resilience, the `ShardArchiveHandler` will utilize a SQLite database to preserve its current state whenever there are active, paused, or queued downloads. The `shard_db` section in the configuration file allows users to specify the location of the database to use for this purpose.
+
+###### SQLite Table Format
+
+| Index | URL                                 |
+|:-----:|:-----------------------------------:|
+| 1     | ht<span />tps://example.com/1.tar.lz4 |
+| 2     | ht<span />tps://example.com/2.tar.lz4 |
+| 5     | ht<span />tps://example.com/5.tar.lz4 |
+
+##### SSLHTTPDownloader
+
+While the archive handler maintains a list of all partial and queued downloads, the `SSLHTTPDownloader` stores the raw bytes of the file currently being downloaded. The partially downloaded file will be represented as one or more `BLOB` entries in a SQLite database. As the maximum size of a `BLOB` entry is currently limited to roughly 2.1 GB, a 5 GB shard file for instance will occupy three database entries upon completion.
+
+###### SQLite Table Format
+
+Since downloads execute serially by design, the entries in this table always correspond to the content of a single file.
+
+| Bytes  | Size       | Part |
+|:------:|:----------:|:----:|
+| 0x...  | 2147483647 | 0    |
+| 0x...  | 2147483647 | 1    |
+| 0x...  | 705032706  | 2    |
+
+##### Config File Entry
+The `download_path` field of the `shard_db` entry will be used to determine where to store the recovery database. If this field is omitted, the `path` field will be used instead.
+
+```dosini
+# This is the persistent datastore for shards. It is important for the health
+# of the ripple network that rippled operators shard as much as practical.
+# NuDB requires SSD storage. Helpful information can be found here
+# https://ripple.com/build/history-sharding
+[shard_db]
+type=NuDB
+path=/var/lib/rippled/db/shards/nudb
+download_path=/var/lib/rippled/db/shards/
+max_size_gb=50
```

Comment:
I think its better to just give some comments explaining these properties. That seems to be the convention as is.

Reply:
I hesitate to agree with the statement that a comment is better than a good name in this situation.  The name is firmly attached to what the user is dealing with.  The comment is not.  That said, there are lots of situations where it's hard to change a name.  I'm asking...

1. Are with in one of those situations, when the name is hard to change?  If not,
2. Are there names that would more explicitly describe _to a user_ (not to you) what they are setting?  If so, use those names.

I'm not gonna push any harder on this particular point.  But I feel that your statement that a comment is better that a good name is short sighted.

Reply:
Let me see how easy this is to change

Reply:
There seems to already be a convention for referring to the db path with the key `path`, so I think we should preserve that. Also, since this key is under a section called `shards_db`, prepending the key seems redundant. How do you feel about just renaming the download path key?

Reply:
Let me see if I can get an expert to weigh in.  I think the following are legit questions, but you should feel free to change the questions however you see fit, @undertome.

@mDuo13, a new field is being added to the `[shard_db]` config stanza.  That stanza already contains a path to where the shard database lives, that's currently called `path=`.

The new field is being added to `[shard_db]` stanza allows a separate path where shard imports occur.  That new field is currently called `download_path=`.

I have a concern that `path=` and `download_path=` are not very descriptive of what the fields actually do.  When there was only one `path=` in the `[shard_db]` config stanza then `path=` was fine.  With two fields I think the risk of confusion is significant.

So I think there are three questions.

1. Shards have been out in the world for a while now, but we've told people not to use them.  Given that, do we feel like the `path=` field cannot change its name?  Or, since we've been telling folks not to use shards, is that name open for change?

2. Are there better names for these fields than the ones that have been selected?

3. Is the flexibility of allowing the import to take place anywhere other than in the `[shard_db]` path useful enough that it justifies the potential confusion?

Thanks for whatever suggestions you may have.

Reply:
A few things worth noting about the `download_path` field and its related download command.

- The `download_path` field is optional and defaults to use `path` if not specified.
- The shard download command is an admin only command that is manually commenced.

Although we have advised users to disable shards in January, they have been in use for nearly two years. Changing the `path` field at this time would likely create more harm than good. There is an opportunity to revise the field name with an upcoming feature that Devon has already begun. That change involves configuring multiple shard store paths and would likely be a good opportunity to consider renaming the field.

> Is the flexibility of allowing the import to take place anywhere other than in the [shard_db] path useful enough that it justifies the potential confusion?

Yes. Allowing a user to configure an optional path for the temporary download directory gives them the flexibility of using a mechanical or non RAID solution as storage. This is important as enterprise SSD storage is very expensive and not necessary for this purpose.

Reply:
Cool.  That upcoming opportunity to revise `path` is useful to know.  So a reasonable way forward would be to come up with the name(s) that we would like to use when the new feature is available.  Once we have that in hand, we can consider a replacement for `download_path` that harmonizes well with the intended new name(s).  We probably leave `path` alone.

Does that sound like a reasonable approach?